Create a converter for amusa from bufr to ioda

xincjin-NOAA opened the pull request:
Amsua has two NCEP bufr files which need to be converted to ioda format and merged together.

The two ncep bufr example files are gdas.t00z.esamua.tm00.bufr_d and gdas.t00z.1bamua.tm00.bufr_d.

1. created one json file for the converter:: bufr2ioda_ncep_amsua.json
2. created two yaml files for the converter: bufr2ioda_combine_ncep_esamua.yml, bufr2ioda_combine_ncep_1bamua_ta.yml
3. fix files are added: amsua_metop-c_v2.ACCoeff.nc [might need more]
4. the following python scripts are added to transfer bufr to ioda format and merge them.
    - antcorr_application.py
    - bufr2ioda_base.py
    - bufr2ioda_combine_ncep_amsua.py
    - utils.py
5. run_bufr2ioda.py is updated for the converter


To test the changes in this PR, we will need to checkout feature/query_python_ext for iodaconv.
https://github.com/JCSDA-internal/ioda-converters/tree/feature/query_python_ext

RussTreadon-NOAA committed:
update prototype_3d observation list to process satwind_goes-16, update obsfile for satwnd.abi_goes-16

RussTreadon-NOAA committed:
Merge develop at 957a2fc into feature/gdas-validation

RussTreadon-NOAA committed:
update GDAS/orion.lua to use spack-stack-1.5.1, bufr/12.0.1, and atlas/0.35.0

emilyhcliu committed:
Remove a double-declaired "obs post filters" in satwnd GOES YAMLs.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Modify BUFR converter and filter yamls for GOES AMV (#724)

Two updates for GOES AMV

1. BUFR Converter:
- remove wind direction and speed
- change sensorZenithAngle to satelliteZenithAngle
2. UFO Filters:
- change `sensorZenithAngle` to `satelliteZenithAngle`
- add `linear obs operator` section
The declaration of `linear obs operator" is necessary for GOES AMV since
there is no linearized component for applying the 10-meter factor.

The increments look reasonable:
```
0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-1.642335e-01 Max:+1.660654e-01 RMS:+2.746685e-04
  0: northward_wind                               | Min:-2.120068e-01 Max:+1.663539e-01 RMS:+2.631295e-04
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------
```

The UFO filter results vs. GSI  -- looks good


![ufo_abi_goes-16_omf_windEastward_qc_time1](https://github.com/NOAA-EMC/GDASApp/assets/36091766/017aeabc-d6ca-4643-b037-16f46dde76a0)


![gsi_abi_goes-16_omf_windEastward_qc_time1](https://github.com/NOAA-EMC/GDASApp/assets/36091766/d99599a2-3f36-4ef6-aa2e-4611500b8c43)

CoryMartin-NOAA committed:
Make the python bufr2ioda driver more generic (#725)

Rather than each new ob type being manually added to the list, we can
use `glob.glob` to get the list of available observation types.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Modify BUFR converter and filter yamls for MetoOp Scatwind (#731)

Two updates for MetOp SCATWIND; one minor updates for GOES AMV 

1. BUFR Converter:
    - remove wind direction and speed
    - add `stationElevation` as zero 

2. UFO Filters:
    - add linear obs operator section
The declaration of `linear obs operator" is necessary for GOES AMV since
there is no
      linearized component for applying the 10-meter factor.
- add `Gaussian Thinning` filter to thin data with 75 km box (equal box
size, prefer data close to central of the box)

3. For GOES AMV YAML files, change output data extension from `nc4` to
`nc`

UFO vs. GSI O-F and QC comparison: O-F and data count passed look close
![ufo_ascat_metop-b_omf_windEastward_qc
(1)](https://github.com/NOAA-EMC/GDASApp/assets/36091766/f234543e-e0b5-4c6e-add9-3c7365c0c469)

![gsi_ascat_metop-b_omf_windEastward_qc](https://github.com/NOAA-EMC/GDASApp/assets/36091766/37f4fba8-bd2a-4a54-b45e-4055e3ef3142)

3Dvar increments are reasonable with MetOp-B SCATWIND alone.
```
0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-1.741092e+00 Max:+1.712287e+00 RMS:+2.282584e-03
  0: northward_wind                               | Min:-1.825957e+00 Max:+1.930385e+00 RMS:+2.496075e-03
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------

```

3DVar increments are also reasonable with MetOp-B + MetOp-A assimilated

CoryMartin-NOAA committed:
Have the driver script process both python scripts and YAMLs + bufr2ioda.x (#730)

This PR allows for all obtypes to be processed in two loops.

We still need to figure out a way to process these in parallel
(somewhat).

CoryMartin-NOAA committed:
Use parallel processing to speed up obs processing (#733)

Using python multiprocessing to generate obs in parallel. The current
list of obs goes from 10+ minutes to completing in ~5.5 minutes.

Commit:
temp

CoryMartin-NOAA committed:
Fix YAML so minimization works for satwind G17 (#743)

@RussTreadon-NOAA noted that `obs linear operator` should be `linear obs
operator` in the GOES-17 satwind YAML. This fixes that.

CoryMartin-NOAA committed:
Consolidate parallel processing (#745)

This reduces runtime further to run everything in one parallel pool.
Now, things complete in under 4 minutes (as of now).

BrettHoover-NOAA committed:
Add AHI/Himawari-8 assimilation to end-to-end GDASApp validation (#746)

Adding satwinds from the Advanced Himawari Imager (AHI) from Himawari-8
to GDASApp end-to-end testing

new files include:
parm/atm/obs/config/satwind_ahi_h8.yaml: QC filter YAML for AHI
Himawari-8 satwinds
parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_ahi.json: JSON containing data
format, sensor, and satellite information for AHI Himawari-8 satwinds
ush/ioda/bufr2ioda/bufr2ioda_satwind_amv_ahi.py: bufr2ioda code for
extracting AHI Himawari-8 satwinds from BUFR

modified files include:
ush/ioda/bufr2ioda/run_bufr2ioda.py: added `"satwind_amv_ahi"` to list
`BUFR_py`

See #741 for details on testing. JEDI/GSI comparisons look good with GSI
thinning turned off, but there are large ob-count disparities both
before and after QC when comparing GSI+thinning to JEDI+thinning. The
tested `Gaussian Thinning` filter is included in the YAML but commented
out with a note.

---------

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>
Co-authored-by: Cory Martin <cory.r.martin@noaa.gov>

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

Commit:
amsua converter initial set

xincjin-NOAA committed:
merge gdas-validation

merge gdas-validation

xincjin-NOAA committed:
forgot undating run_bufr2ioda.py

xincjin-NOAA committed:
a working version

CoryMartin-NOAA committed:
Write the bufr2ioda config files in DATA (#748)

Instead of writing the filled out `.json` and `.yaml` files to
`$COM_OBS`, we write them to `$DATA` so that they get purged unless
`KEEPDATA="YES"`

xincjin-NOAA committed:
remove a json file

xincjin-NOAA committed:
Merge branch 'feature/gdas-validation' into feature/gdas-validation_amsua

xincjin-NOAA committed:
correct code norm

xincjin-NOAA committed:
a few more

xincjin-NOAA committed:
a few more

xincjin-NOAA committed:
a few more

xincjin-NOAA committed:
a few more

xincjin-NOAA committed:
dele last line

xincjin-NOAA committed:
no new line at the end of the file

emilyhcliu commented:
@xincjin-NOAA I converted this PR to a draft PR because we need to checkout feature/query_python_ext for testing.

XuanliLi-NOAA committed:
Update GNSSRO assimilation to end-to-end GDASApp validation (#753)

Updating GNSSRO converter and adding the yaml file to GDASApp end-to-end
testing based on [#712](https://github.com/NOAA-EMC/GDASApp/pull/712).

New file:
parm/atm/obs/config/gnssro.yaml: Added QC filters in the YAML

Updated files:
parm/ioda/bufr2ioda/bufr2ioda_gnssro_bufr.json: Updated the satellite
information
ush/ioda/bufr2ioda/bufr2ioda_gnssro_bufr.py: Updated pccf, qfro, and
satelliteAscending flag

Refer to [issue #750](https://github.com/NOAA-EMC/GDASApp/issues/750)
for testing details. JEDI/GSI comparisons revealed a cutoff near
obs=0.03 Rad due to super refraction 2. The number of observations after
QC in JEDI is also inconsistent with GSI.

emilyhcliu committed:
Update for ATMS for End-to-End Test (#757)

List of items to be added in this PR:
- BUFR Converter: Add YAML for converting ATMS BUFR to IODA
   - The ObsValue is antenna temperature
   - Input BUFR is normal feel 

- Testing YAML: re-evaluation results due to update CRTM from v2.3.0 to
v2.4.1-jedi.1
   - atms_npp_noqc.yaml --- 100% replication
   - atms_npp.yaml  --- 100 % replication
   - atms_n20_noqc.yaml --- 100 % replication
- atms_n20.yaml --- ~ 1100% replication: two channel 7 observations
difference
  
- Config YAML: 
* add two read routine QC: data thinning and removal of data from scan
edge
   - atms_npp.yaml
   - atms_n20.yaml   

- Paring PR in JCSDA UFO repository
- This PR also has a [paring PR in JCSDA UFO
repository](https://github.com/JCSDA-internal/ufo/pull/3094).
   - It is OK to merge this one into GDASApp before the paring UFO PR
- The related test data (obs and geoval files from GSI) for UFO
Evaluation have been updated (due to CRTM-2.4.1 update) on ORION and
HERA.


Notes:
- End-to-end testing completed without problem. 
- However, the filter results show one of the filters, 88-165 GHz
scattering index check, tossed more data compared to GSI.
- The excess screening from the 88-165 GHz scattering index check is
under investigation and will be reported separately.
- Please check comments below for validation results

emilyhcliu committed:
Update amsua obs YAML files

emilyhcliu committed:
For end-to-end configuration AMSU-A YAML files: update n19 and add n18, n15, metop-b and metop-c

emilyhcliu committed:
Add various platforms for amsua for 3dvar test

emilyhcliu committed:
Bug fixes: define Sensor_ID, modify obsfile names to follow naming convention.

emilyhcliu committed:
Update diagnostic flags, add thinning and scan edge removal filters

emilyhcliu committed:
Remove extra comments

CoryMartin-NOAA commented on CMakeLists.txt:
```diff
@@ -107,7 +107,7 @@ if(BUILD_GDASBUNDLE)
 # Build IODA converters
   option(BUILD_IODA_CONVERTERS "Build IODA Converters" ON)
   if(BUILD_IODA_CONVERTERS)
-    ecbuild_bundle( PROJECT iodaconv GIT "https://github.com/JCSDA-internal/ioda-converters.git" BRANCH develop )
```
note to remove this change before merging

CoryMartin-NOAA replied:
Can we just revert this file so that it is not included in the PR?

xincjin-NOAA replied:
I just did this to revert it:

 git checkout 3f2b79d -- CMakeLists.txt
git commit -m 'revert CMakeLists.txt'
git push

anything else I need to do in order to revert it?

CoryMartin-NOAA replied:
nope, that's perfect, thanks! Note the 3dvar YAML will also need this done

xincjin-NOAA replied:
Thanks!

CoryMartin-NOAA commented on parm/atm/variational/3dvar_dripcg.yaml:
```diff
@@ -33,7 +33,7 @@ variational:
   minimizer:
     algorithm: DRIPCG
   iterations:
-  - ninner: 2
```
This file shouldn't be included/changed in this PR. My guess this is a `git commit -a` case gone awry...

xincjin-NOAA replied:
reverted this. Thanks!

CoryMartin-NOAA commented on parm/ioda/bufr2ioda/amsua_metop-c_v2.ACCoeff.nc:
I'm not really a fan of putting binary files (including netCDF) in this repo. @emilyhcliu can we put this somewhere else?

CoryMartin-NOAA commented on parm/ioda/bufr2ioda/bufr2ioda_combine_ncep_1bamua_ta.yml:
we should use `.yaml` not `.yml` to be consistent with the rest of the repo

CoryMartin-NOAA commented on parm/ioda/bufr2ioda/bufr2ioda_combine_ncep_1bamua_ta.yml:
```diff
@@ -0,0 +1,190 @@
+
+    ioda:
+      backend: netcdf
+      obsdataout: "{{ COM_OBS }}/{{ RUN }}.t{{ cyc }}z.amsua.$(splitvar)_ta.tm00.nc"
```
I think we decided we wanted `amsua_$(splitvar)`?

xincjin-NOAA replied:
The outputs from this are not the final ioda files. We reserve the correct format for the final files.

CoryMartin-NOAA commented on parm/ioda/bufr2ioda/bufr2ioda_combine_ncep_esamua.yml:
```diff
@@ -0,0 +1,187 @@
+
+    ioda:
+      backend: netcdf
+      obsdataout: "{{ COM_OBS }}/{{ RUN }}.t{{ cyc }}z.esamua.$(splitvar).tm00.nc"      
```
same comment as above

xincjin-NOAA replied:
see reply as above

CoryMartin-NOAA commented on parm/ioda/bufr2ioda/bufr2ioda_ncep_amsua.json:
```diff
@@ -0,0 +1,19 @@
+                          { "satellite_name": "GOES-17", "satellite_full_name": "Geostationary Operational Satellite - 17", "satellite_id": 271 }, 
+                          { "satellite_name": "GOES-18", "satellite_full_name": "Geostationary Operational Satellite - 18", "satellite_id": 272 } 
+                       ],
+  "yaml_file"        : ["{{ COM_OBS }}/ncep_esamua_{{ current_cycle | to_YMDH }}.yaml", 
```
I don't think we want the YAML file in `$COM_OBS` but rather `$DATA

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/bufr2ioda_base.py:
```diff
@@ -0,0 +1,67 @@
+
+    def get_yaml_config(self, config):
+        # TODO  add and/or modify ymal contents from config json file. e.g. file names, etc.
+        # What does en_bufr2ioda_yaml.py do?   connection to this?
```
yes, I think we can use gen_bufr2ioda_yaml here

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.py:
```diff
@@ -0,0 +1,94 @@
+R1000 = 1000.0
+R1000000 = 1000000.0
+
+SPLIT = ('metop-a', 'metop-b', 'metop-c', 'n17', 'n18', 'n19')
```
any reason why these need coded in this py file and not a config JSON or YAML?

xincjin-NOAA replied:
removed and get it from the JSON file

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.py:
```diff
@@ -0,0 +1,94 @@
+    obs_Path_prefix = os.path.commonprefix([amsua_files[0][0], amsua_files[1][0]])
+    for ioda_file_0, ioda_file_1, ioda_file_t in ioda_files:
+        try:
+            nc_merge(obs_Path_prefix + ioda_file_0, obs_Path_prefix + ioda_file_1, obs_Path_prefix + ioda_file_t)
```
use `os.path.join`

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.py:
```diff
@@ -0,0 +1,94 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    # parser.add_argument('-t', '--type', type=str, help='Input Satellite type a/e', required=True)
```
why is this commented out?

xincjin-NOAA replied:
The code structure is changed. removed.

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.py:
```diff
@@ -0,0 +1,94 @@
+        print(amsua_files)
+        logger.info('Converting amsua type {} done!'.format(sat_type))
+
+    print('--start to merge--')
```
logger instead of print unless this is just for debugging

xincjin-NOAA replied:
replaced print

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/run_bufr2ioda.py:
```diff
@@ -52,8 +52,11 @@ def bufr2ioda(current_cycle, RUN, DMPDIR, config_template_dir, COM_OBS):
-    shutil.copy(os.path.join(config_template_dir, "atms_beamwidth.txt"),
-                os.path.join(DATA, "atms_beamwidth.txt"))
+    fix_files = ["atms_beamwidth.txt",
+                 "amsua_metop-c_v2.ACCoeff.nc"
```
need to find a place for this .nc file...

xincjin-NOAA replied:
put a note in the code

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/run_bufr2ioda.py:
```diff
@@ -103,6 +106,25 @@ def bufr2ioda(current_cycle, RUN, DMPDIR, config_template_dir, COM_OBS):
         # if os.path.exists(yaml_output_file):
         #     rm_p(yaml_output_file)
 
+        # Specify observation types to be processed by a script with combined methods
```
indentation of comment should match code it is referring to

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/run_bufr2ioda.py:
```diff
@@ -103,6 +106,25 @@ def bufr2ioda(current_cycle, RUN, DMPDIR, config_template_dir, COM_OBS):
+
+    for obtype in BUFR_combine:
+        logger.info(f"Convert {obtype}...")
+        json_output_file = os.path.join(COM_OBS, f"{obtype}_{datetime_to_YMDH(current_cycle)}.json")
```
$DATA, not $COM_OBS (I believe)

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/utils.py:
```diff
@@ -0,0 +1,87 @@
+import time
+
+
+def nc_merge(file_name1, file_name2, target_name):
```
This is fine for now, but feel strongly this is something that a compiled IODA OOPS app should do.

CoryMartin-NOAA commented on ush/ioda/bufr2ioda/utils.py:
```diff
@@ -0,0 +1,87 @@
+                    ncf_var.setncattr(attr, var.getncattr(attr))
+
+
+def timing_decorator(func):
```
Pass in a logger and use that instead of print?

xincjin-NOAA replied:
replaced print with a logger.

BrettHoover-NOAA committed:
[End-to-End Test Code Sprint] Add SEVIRI METEOSAT-8 satwinds to end-to-end testing (#764)

Adding satwinds from the Spinning Enhanced Visible and InfraRed Imager
(SEVIRI) from METEOSAT-8 to GDASApp end-to-end testing

new files include:
parm/atm/obs/config/satwind_seviri_m8.yaml: QC filter YAML for SEVIRI
METEOSAT-8 satwinds
parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.json: JSON containing
data format, sensor, and satellite information for SEVIRI METEOSAT-8
satwinds
ush/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.py: bufr2ioda code for
extracting SEVIRI METEOSAT-8 satwinds from BUFR

End-to-end testing results are provided in #758 - Note: Thinning filter
is provided in YAML but turned off due to irreconcilable differences
with GSI in testing.

---------

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>

emilyhcliu committed:
Do nothing.

emilyhcliu committed:
Merge branch 'feature/gdas-validation' into feature/gdas-validation-amsua

emilyhcliu committed:
Merge branch 'feature/gdas-validation' into feature/gdas-validation_amsua

BrettHoover-NOAA committed:
Added SEVIRI/METEOSAT-11 satwinds to end-to-end validation (#767)

Small code-change to include SEVIRI/METEOSAT-11 satwinds in end-to-end
validation. This uses the hook-ins already provided by the
SEVIRI/METEOSAT-8 satwinds code that was merged in
https://github.com/NOAA-EMC/GDASApp/pull/764.

New file: parm/atm/obs/config/satwind_seviri_m11.yaml - this is nearly a
direct copy of parm/atm/obs/config/satwind_seviri_m8.yaml, since they
utilize the same filters. Only the header information has changed.

Modified file: parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_seviri.json -
METEOSAT-11 was added to the `satellite_info` group

End-to-end testing looks very similar to what was seen for
SEVIRI/METEOSAT-8 satwinds, detailed in
https://github.com/NOAA-EMC/GDASApp/issues/758. A brief run-down:

(LW)IR winds are in QC agreement down to a 0.02% difference
(cloud-top)WV winds are in QC agreement down to a 0.20% difference
VIS winds differ by 4.5%, but there are only 22 accepted winds in GSI
and 23 in JEDI, the difference is a single satwind

All ob, HofX, and ob-minus-HofX differences look good

Increment summary, from gdasatmanlrun.log:
```
  0: ----------------------------------------------------------------------------------------------------
  0: Increment print | number of fields = 8 | cube sphere face size: C768
  0: eastward_wind                                | Min:-3.479765e-01 Max:+3.430433e-01 RMS:+4.990993e-04
  0: northward_wind                               | Min:-4.226316e-01 Max:+3.376561e-01 RMS:+4.889429e-04
  0: air_temperature                              | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: surface_pressure                             | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: specific_humidity                            | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: cloud_liquid_ice                             | Min:+0.000000e+00 Max:+1.618770e-20 RMS:+1.293217e-23
  0: cloud_liquid_water                           | Min:+0.000000e+00 Max:+1.474788e-19 RMS:+2.167418e-22
  0: ozone_mass_mixing_ratio                      | Min:+0.000000e+00 Max:+0.000000e+00 RMS:+0.000000e+00
  0: ----------------------------------------------------------------------------------------------------
```

Co-authored-by: Brett Hoover <bhoover@Orion-login-1.HPC.MsState.Edu>

CoryMartin-NOAA committed:
Merge branch 'feature/gdas-validation' into feature/gdas-validation-amsua

emilyhcliu committed:
A few updates for ATMS end-to-end testing (#768)

In [ previous PR](https://github.com/NOAA-EMC/GDASApp/pull/757) for
ATMS, the end-to-end testing was done without data thinning. And the
excessive QC filtering in clear-sky areas, compared to GSI, was found.
(See
[plots](https://github.com/NOAA-EMC/GDASApp/pull/757#issuecomment-1819753244))

This PR includes two fixes for ATMS in the end-to-end testing and one
new feature:
1. fix the excess QC filtering in clear-sky areas. ---> This has a
paring [UFO PR #3122](https://github.com/JCSDA-internal/ufo/pull/3122)
2. add data thinning 
3. add the diagnostic flags (QC) --- **This can [reproduce QC
flags](https://github.com/NOAA-EMC/GDASApp/pull/768#issuecomment-1828209921)
from GSI in UFO**

**To test updates in this PR, please check out the UFO branch:
[feature/satrad](https://github.com/JCSDA-internal/ufo/tree/feature/satrad)
from JCSDA-internal in gdas-validation. This branch consolidates all
proposed code changes to UFO for gdas validation:
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3122
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3121
UFO PR #https://github.com/JCSDA-internal/ufo/pull/3094

---------

Co-authored-by: Cory Martin <cory.r.martin@noaa.gov>

emilyhcliu committed:
Merge branch 'feature/gdas-validation' into feature/gdas-validation-amsua

emilyhcliu committed:
Remove temporary checkout from JCSDA-internal/ufo

xincjin-NOAA committed:
add no python option

emilyhcliu committed:
Re-arrange and tidy up the YAML files for AMSU-A

emilyhcliu committed:
change post filters to pre filters for diagnostic flags

CoryMartin-NOAA committed:
Revert 3dvar_dripcg.yaml

xincjin-NOAA committed:
a new version of amsua converter

xincjin-NOAA committed:
Merge branch 'feature/gdas-validation_amsua' of https://github.com/NOAA-EMC/GDASApp into feature/gdas-validation_amsua

xincjin-NOAA commented:
pushed a new version in which the branch of feature/query_python_ext for iodaconv is not needed.

xincjin-NOAA committed:
revert CMakeLists.txt

xincjin-NOAA committed:
revert parm/atm/variational/3dvar_dripcg.yaml

xincjin-NOAA committed:
Merge branch 'feature/gdas-validation-amsua' into feature/gdas-validation_amsua_test

merge together

xincjin-NOAA committed:
for test

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

xincjin-NOAA committed:
a few small changes

xincjin-NOAA committed:
optimized apply_ant_corr

emilyhcliu committed:
Minor observation YAML updates for ATMS, SATWIND and SCATWIND (#789)

This PR includes the following:
1. YAML name change - add sensor (ABI) to the filename of observation
YAML for SATWIND from GOES platforms
2. YAML name change - add sensor (ASCAT) to the filename of observation
YAML for SCATWIND from MetOp platforms
3. update the YAML name changes in the list for 3dvar 
4. Add `CO2` in obs forward operator and add `linear obs operator`
without CO2 for ATMS

These changes have no impact on 3DVar results.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

xincjin-NOAA committed:
add remove function

emilyhcliu committed:
Update IASI  YAML files for end-to-end testing (#769)

RussTreadon-NOAA commented:
Attempt to test changes in PR #763 and this PR together on Orion.   `gdasprepatmiodaobs` fails with 
```
^[[38;21m2023-12-05 17:49:18,877 - INFO     - run_bufr2ioda.py: Convert combine_ncep_amsua...^[[0m
^[[38;21m2023-12-05 17:49:18,877 - INFO     - gen_bufr2ioda_json.py: Using /work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/parm/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.json as input^[[0m
Traceback (most recent call last):
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/run_bufr2ioda.py", line 148, in <module>
    bufr2ioda(args.current_cycle, args.RUN, args.DMPDIR, args.config_template_dir, args.COM_OBS)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/run_bufr2ioda.py", line 74, in bufr2ioda
    gen_bufr_json(config, template, json_output_file)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/sorc/gdas.cd/ush/ioda/bufr2ioda/gen_bufr2ioda_json.py", line 19, in gen_bufr_json
    bufr_config = parse_j2yaml(template, config)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/python/wxflow/yaml_file.py", line 176, in parse_j2yaml
    yaml_dict = YAMLFile(data=yaml_file)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/python/wxflow/yaml_file.py", line 35, in __init__
    self.update(config)
  File "/work2/noaa/da/rtreadon/gdas-validation/global-workflow/ush/python/wxflow/attrdict.py", line 120, in update
    other.update(args[0])
ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

File `parm/ioda/bufr2ioda/bufr2ioda_combine_ncep_amsua.json` does not exist in either PR #763 nor this PR.   This PR only contains one `json` file, `parm/ioda/bufr2ioda/bufr2ioda_ncep_amsua.json`

Is there an error in `run_bufr2ioda.py` when it forms the name of the file to process?  Is the file we want actually `parm/ioda/bufr2ioda/bufr2ioda_ncep_amsua.json`?

I'm stuck and don't know what to try next.  Maybe I'm getting ahead of myself since this PR is still in _draft_ mode.

xincjin-NOAA commented:
@RussTreadon-NOAA, Sorry for the little slow progress. You can checkout the branch of feature/gdas-validation_amsua_test.
Or should I create a new PR and or issue?  In that new branch I have merged above two together

RussTreadon-NOAA commented:
Since your PR is goes with Emily's PR #763, please coordinate with @emilyhcliu  as to the best way to move forward.  

If we wind up with a single branch with all the amsua changes, I can check out that branch and see what happens when I run it using the global-workflow.

xincjin-NOAA committed:
temp

xincjin-NOAA committed:
merge

xincjin-NOAA committed:
a few small changes

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/gdas-validation

emilyhcliu committed:
Update observation YAMLs for conventional surface pressure (#803)

This PR includes the following updates for conventional surface pressure
1. Add observation YAML under configuration
2. Update observation YAML under testing
3. Add conv_ps to 3dvar obs list
4. Set window shift to true (time boundary is inclusive)
5. Validate 3dvar for surface (station) pressure (end-to-end)
- [QC
validation](https://github.com/NOAA-EMC/GDASApp/pull/803#issuecomment-1851387894)
-
[Increments](https://github.com/NOAA-EMC/GDASApp/pull/803#issuecomment-1852496937)

See [PR #792 ](https://github.com/NOAA-EMC/GDASApp/pull/792) for the
BUFR converter work to convert the following surface pressure
observation into IODA:
- SFC - 181, 187
- SFCSHIP - 180, 183
- SONDE - 120

The related UFO evaluation (using GSI obs and geovals) results for
surface pressure are documented
[here](https://docs.google.com/presentation/d/1OvVMYOutxVyk0GE93Jzoh3hQVoj-dCgrwsVWDDlPT1s/edit?usp=sharing).

xincjin-NOAA committed:
a working version matched with GSI

xincjin-NOAA committed:
code norm check

xincjin-NOAA committed:
for ctest

xincjin-NOAA committed:
gdas_validation merge

xincjin-NOAA committed:
merge develop

xincjin-NOAA commented:
When I merge the develop into this branch I got these two message, I merged them manually so that the branch passed ctest. I am not sure if my merge is correct or not:  
Auto-merging parm/atm/variational/3dvar_drpcg.yaml
CONFLICT (content): Merge conflict in parm/atm/variational/3dvar_drpcg.yaml
Removing ci/stable_mark.sh
Auto-merging CMakeLists.txt
CONFLICT (content): Merge conflict in CMakeLists.txt
Automatic merge failed; fix conflicts and then commit the result.

xincjin-NOAA commented:
![ioda_amsua_n19_GSI_IODA_1_ObsValue](https://github.com/NOAA-EMC/GDASApp/assets/122945910/e749b5cf-f3e6-4ba3-83da-f4a3890b8477)

![ioda_amsua_n19_IODA_1_ObsValue](https://github.com/NOAA-EMC/GDASApp/assets/122945910/2c95ab9b-52c8-44da-972c-f2ab44cf752d)

The comparison of the obs values from GSI and IODA are agreed well

xincjin-NOAA commented:
![ioda_amsua_n19_GSI_IODA_1_GsiHofXBc](https://github.com/NOAA-EMC/GDASApp/assets/122945910/90501993-69f2-41a0-a124-b21f4adfab4e)

![ioda_amsua_n19_IODA_1_hofx0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/5b54c7d7-53f3-412a-85a3-6217c961b8ab)

These two are for hofx0

emilyhcliu commented:
Change the base from feature/gdas-validation to develop.
Now, the number of file changes is down from 235 to 40

xincjin-NOAA commented:
**Figure 1**
Here are the channel statistics plots (#obs, bias/std for OMF and final obs error) for JEDI and GSI runs. Note: channels 7 and 8 are excluded in JEDI.

**O-F Bias/Standard Deviation**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/856daa48-5c1c-4dcc-9ed6-a14fd6066f9b) | ![JEDI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/76cfc385-4653-43e8-a6e6-522acbd48907)

**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/e9001e3b-054c-4c6a-a470-209e1facffe5) | ![JEDI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/debae5f6-9073-46d7-8471-4da8805d0c14)

**Numbers PassedQC**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/3987d5b6-8fc4-4266-8597-2b7d78caca37) | ![JEDI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/9abfcebb-6268-4edf-8dc3-d86b58ae02c8)

ADCollard commented:
@xincjin-NOAA Thanks for this!   The OmF and final obs error for channel 3 seems a little higher for JEDI (it looks like it might just be a few points).   Do we know why this is?

xincjin-NOAA commented:
**Figure 2 **
Maps for Channel 3 to justify the differences between GSI and JEDI above.

**O - F**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_ObsValue_GsiHofXBc](https://github.com/NOAA-EMC/GDASApp/assets/122945910/43aea5c8-43a2-4981-aa19-2aa3694bb8b9) | ![ioda_amsua_n19_IODA_3_ObsValue_hofx0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/012f7e62-197f-4238-bd42-1966e41d1a81)


**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_GsiFinalObsError](https://github.com/NOAA-EMC/GDASApp/assets/122945910/d784bbad-85a7-4348-91dc-88e2019d934a)| ![ioda_amsua_n19_IODA_3_EffectiveError0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/f96cb70d-8a98-4d5a-bfb7-724bf139195b)

RussTreadon-NOAA commented:
Please update `feature/gdas-validation_amsua` to be consistent with the head of GDASApp `develop`.  Note that the following files 
```
parm/atm/obs/config/amsua_n19.yaml
parm/atm/obs/lists/gdas_prototype_3d.yaml
parm/atm/variational/3dvar_drpcg.yaml
```
conflict with `develop`.

xincjin-NOAA commented:
I am going to merge the develop into this branch and make a few final test runs.

xincjin-NOAA committed:
merge with develop

emilyhcliu commented:
The merge is a bit odd and tricky.
The AMSU-A yaml files under `parm/atm/obs/config`  in your branch before merging with the `develop` are new yaml files except for amsua_n19.  So, when you merge your branch with the develop (yaml file has j2 extension under config), you will get a conflict for amsua_n19.yaml

The amsua_n19.yaml.j2 in the `develop` is old.  Could you confirm that the amsua_n19.yaml.j2 in the merged branch has all the updates (e.g. thinning and some changes in the filters) from your branch?
Also, it looks like you will need to manually add j2 extension for the merged amsua yaml files under config.

xincjin-NOAA commented:
@emilyhcliu I will check those and make a test run then.

Thanks,

xincjin-NOAA committed:
make it working after merging develop

xincjin-NOAA commented:
Merged the develop into this branch and tested on orion.

xincjin-NOAA commented:
@ADCollard and @emilyhcliu , After a small modifying of QC filters on the GSI run  the OmF and final ObsError difference between GSI and JEDI become smaller(the plots were updated), though there still a little differences. We also plot the maps of them and don't find any odd behaviors.  We conclude that the differences are caused mainly by the difference of filters.

Not sure if we still want to make more tests to make  the differences smaller.

ADCollard commented:
@xincjin-NOAA Can you confirm that the plots have been updated (channel and maps plots)?   They say they were most recently edited last week.

Having said that they do look a little better than I remember.   Do you have the unthinned versions of these plots?

Can you also explain what was changed in the GSI to improve the agreement?

Thanks.

xincjin-NOAA commented:
@ADCollard Yes I updated the plots (might be it is better just add new plots later).  @emilyhcliu ran the new GSI with some changes in the filters.  I am going to see if I can find the results from the unthinned runs.

xincjin-NOAA commented:
**Figure 3**

**Obs comparisons between those from GSI and IODA(after converter and not thinned) **

GSI | JEDI |
:-------------------------:|:-------------------------:
![image](https://github.com/NOAA-EMC/GDASApp/assets/122945910/b1c3a352-f09c-43ed-99e7-e718e5040547) | ![image](https://github.com/NOAA-EMC/GDASApp/assets/122945910/f0c69aa8-3b4e-4d19-8de4-6a3558893e66)


GSI | JEDI |
:-------------------------:|:-------------------------:
![image](https://github.com/NOAA-EMC/GDASApp/assets/122945910/d1ed2dda-c381-44e2-885c-0c21f86dc6b5) | ![image](https://github.com/NOAA-EMC/GDASApp/assets/122945910/dd0e0ae5-a53a-4be8-812f-8761c428f5e9)

**Conclusion**
The Obs are almost the same after convert to ioda format.

ADCollard commented:
@xincjin-NOAA Can you plot O-B after QC?  Thanks.

xincjin-NOAA commented:
@ADCollard These are from the files directly converted from the bufr file and no QC yet in case of IODA.  Is O-F the same as O-B? The O-F plots above are after QC.

ADCollard commented:
@xincjin-NOAA Yes O-F is the same as O-B.   I thought you had produced the O-B plots after QC but before thinning? Maybe @emilyhcliu did?

emilyhcliu commented:
@xincjin-NOAA Are you now running the end-to-end with the latest T2O fix?  If so, then your result from GADS is high resolution (C768).  Please confirm if you are running end-to-end in C768 resolution.  Check your `config.base` and search `CASE`.  
Let me know which resolution you are using.  I will point you to the correct IODA file from GSI.

emilyhcliu commented:
@xincjin-NOAA I checked out your branch and will test them today.

xincjin-NOAA commented:
**Figure 4**

Here are the channel statistics plots for results without thinning and with QC. ( the amsua n19 file is merged from two files) 

**O-F Bias/Standard Deviation**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/b0471484-c60b-4f2c-b305-7437bde0477c)| ![JEDI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/6ba1a3c4-291c-4128-a627-c603806403b0)

**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/99bb9cd9-6e22-45e4-8741-19d16c87f451)| ![JEDI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/5b888ad6-3ee5-404f-a522-fc7973f0edf5)


**Numbers PassedQC**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/62bee9e4-b72e-4150-aa47-c97ef9ad830f)| ![JEDI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/7af63be7-e56c-4a1a-ade6-5b63e2c6a53c)




Maps for Channel 3 .

**O - F**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_ObsValue_GsiHofXBc](https://github.com/NOAA-EMC/GDASApp/assets/122945910/a22745a7-4603-4254-be30-0e9c1e684425)| ![ioda_amsua_n19_IODA_3_ObsValue_hofx0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/c1b673d4-9bd4-4833-a60f-7812b78cbef2)

**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_GsiFinalObsError](https://github.com/NOAA-EMC/GDASApp/assets/122945910/3591ca8b-7c12-49b5-99f3-878ecc2f5ba8)| ![ioda_amsua_n19_IODA_3_EffectiveError0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/454fe3d2-62ac-485f-9310-2dd8f0d34b14)


There are no large differences found.

xincjin-NOAA commented:
**Figure 5**

Here are the channel statistics plots for results without thinning and after QC for a single 1bamsua file.

**O-F Bias/Standard Deviation**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/2a87b326-4c62-4207-818d-4ce01d796fee)| ![JEDI_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/06196eed-5886-4870-9347-87d98baff09c)


**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/4af6ff82-89a3-4571-9d65-4a3709993693)| ![JEDI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/3521c813-32eb-42ec-8d07-ab4ac8c1b850)

**Numbers PassedQC**
GSI | JEDI |
:-------------------------:|:-------------------------:
![GSI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/902dc2eb-a459-4495-882d-ec919db26f06)| ![JEDI_PassedQC_Countsamsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/03d97690-7d20-4094-9d6d-0af2a081020a)


Maps for Channel 3 .

**O - F**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_ObsValue_GsiHofXBc](https://github.com/NOAA-EMC/GDASApp/assets/122945910/1b69331c-02bf-4c80-a528-c438a797b0d5)| ![ioda_amsua_n19_IODA_3_ObsValue_hofx0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/5f692f17-5ba9-4060-9160-e67615a26ce7)

**Final ObsError**
GSI | JEDI |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_3_GsiFinalObsError](https://github.com/NOAA-EMC/GDASApp/assets/122945910/a7ec1c15-0142-4ace-89d2-ba1865357585)| ![ioda_amsua_n19_IODA_3_EffectiveError0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/aa2aeb07-6b1d-4592-b737-7492b2e1b15a)


They agree well with each other in general.

xincjin-NOAA commented:
**Summary**

*Figure 1* , the channel statistics plots (#obs, bias/std for OMF and final obs error) for JEDI and GSI runs show general agreement between GSI and JEDI results, except for the channel 3 which show some differences in O-F and Final-error. In order to understand this we explored different causes as follow:   

- The maps for channel 3 (*Figure 2*) show that there are not any place where significant differences existed.
- Figure 3 demonstrated that there are no differences between GSI and JEDI after converted to IODA data format.
- Figure 4, the plots from results without thinning and after QC, show no large differences
- Figure 5, the plots which is drawn based on the results with a single amsua file without merging, also show no large differences, though some differences are existed here or there.

Therefore, we can conclude that the small differences come from different methods used in GSI and JEDI in terms of thinning, different QC filters in general. The difference identified in *Figure 1* for channel 3 is from different thinning methods between GSI and JEDI.

xincjin-NOAA commented:
There are two concerns left from this PR:

- find a place  to store the Antenna Coefficient NetCDF files
- merge two .nc files is okay for now, but it is better to implement it in the ObSpace 

create another issue for this: #987

emilyhcliu commented:
@xincjin-NOAA  The result for the unthinned case looks OK in general.  The obs errors for low-peaking channels 1-2 have slightly lower final obs error standard deviation values.   I will run end-to-end with your branch to check the initial/final them obs error and compared with GSI.  I am a bit curious about the slight difference for channels 1-2.  I will post the results of my test here.  

One more comment:  
Please add error handling in the Python converter for missing input files or missing subsets
     - the need to add the error handling was identified by @RussTreadon-NOAA  (see [PR #937](https://github.com/NOAA-EMC/GDASApp/pull/937) for more infomation)
     - here is the example of error handling added in the converter
    
![Screenshot 2024-03-21 at 11 34 43 PM](https://github.com/NOAA-EMC/GDASApp/assets/36091766/52531e2e-818f-470c-9d95-b47af758d9d6)

emilyhcliu commented:
> There are two concerns left from this PR:
> 
> * find a place for *.nc files
> * merge two .nc files
> 
> create another issue for this: #987

Your amsua python code can merge the 1bamua and esamua, isn't it?

For your first concern, do you mean we need to find a place to store the Antenna Coefficient NetCDF files?

xincjin-NOAA commented:
@emilyhcliu 

For the first, Yes.  @CoryMartin-NOAA commented that it is better to merge the two .nc files in Obspace, though it is okay for now.

I will put details there.

xincjin-NOAA commented:
After some experiments by turn off  different combinations of filters and found that the step 10 which is the filter of Grosscheck filter might play a key role in the differences found. I am going to explore the map of them to see if we can find something.

**Figure 6**

**Final ObsError**
Before  | After  |
:-------------------------:|:-------------------------:
![JEDI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/e3b8d5b0-8e90-4d51-be91-7da09006b09e)| ![JEDI_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/e0bdee93-fb84-4e62-8ada-c15c92ff8066)

xincjin-NOAA commented:
**Numbers in Gross Check**
Channel   |1  | 2 | 3 | 4 |5 |6 |7 |8 |9 |10| 11 |12 |13 |14 |15 
:-------:|:----:|:----------:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:|:----:
IODA | 165 | 164  | 54  |  9 |  56 |  27 | 0|0 | 163 |  24 |  28 |  76 | 157 |  54 |39
GSI | 148 | 130 | 59| 8| 62 |40 |0 |0 |156 |18| 29| 69 |149|52| 36

The numbers in channel 1, 2, are much larger than that for channel 3. This can explain the feature noticed above.

**Figure 7**

**Map for Gross Check**
GSI  | JEDI  |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_gsi_3_gross_check](https://github.com/NOAA-EMC/GDASApp/assets/122945910/f2826133-6dc2-4110-8f25-d4fc55447009)| ![ioda_amsua_n19_ioda_3_gross_check](https://github.com/NOAA-EMC/GDASApp/assets/122945910/a61154ba-d982-4409-825f-ea61feb294da)


Gross filter are quite different between GSI and IODA.

xincjin-NOAA commented:
**Figure 8**  Comparisons of the Initial ObsError between GSI and JEDI.

**Initial ObsError**
GSI  | JEDI  |
:-------------------------:|:-------------------------:
![GSI_input_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/1e9bcb6c-d238-4ff1-9bda-87fb581d4cfb)| ![JEDI_Derived_ObsError_ChannelStates_amsua_19_diag_2021080100_amsua_19](https://github.com/NOAA-EMC/GDASApp/assets/122945910/4be88ace-ec9c-4027-a7f5-deb578d66795)

xincjin-NOAA commented:
**Figure 9**  Map comparisons of the Initial ObsError, retrieved clw, and hofx0 between GSI and JEDI (un-thinned data, Channel 1)

**Initial ObsError**
GSI  | JEDI  |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_1_GsiInputObsError](https://github.com/NOAA-EMC/GDASApp/assets/122945910/308eb253-5e1f-44c9-b671-2b7ed62f4b22)| ![ioda_amsua_n19_IODA_1_InitialObsError](https://github.com/NOAA-EMC/GDASApp/assets/122945910/9d8a98eb-7af5-47f9-b9b9-3a27e121eef1)


**Retrieved CLW**
GSI  | JEDI  |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_1_cloudWaterRetrievedFromSimulatedObservation](https://github.com/NOAA-EMC/GDASApp/assets/122945910/454614d1-b42a-4f92-9d86-4b561dec434f)| ![ioda_amsua_n19_IODA_2_CLWRetFromBkg](https://github.com/NOAA-EMC/GDASApp/assets/122945910/835b3326-4084-4f2c-9fd7-dbce1059790e)

**Hofx0**
GSI  | JEDI  |
:-------------------------:|:-------------------------:
![ioda_amsua_n19_GSI_IODA_1_GsiHofXBc](https://github.com/NOAA-EMC/GDASApp/assets/122945910/40480737-5312-4469-96e1-325963b50bff)| ![ioda_amsua_n19_IODA_1_hofx0](https://github.com/NOAA-EMC/GDASApp/assets/122945910/378af2a2-ca20-4f63-9620-60c926cf2ce5)

xincjin-NOAA committed:
some clean and log

xincjin-NOAA commented:
Figures 8 and 9 showed un-thinned statics of initial errors and related retrieved clw and hofx0. 

1. Figure 8 show the initial errors agree well between GSI and JEDI with a little higher in GSI for channels 2 and 3 as noticed by @emilyhcliu above.
2. From the map in Figure 9 we can see a little difference between GSI and JEDI in the initial error and related clw and hofx0

In summary, we think these differences, (mean and std differences are 1%, and 4%, respectively) are acceptable considering there are many differences between two systems.  Any comments from @emilyhcliu and @ADCollard?

xincjin-NOAA commented:
@emilyhcliu error handling in the codes were checked and added some more.

CoryMartin-NOAA commented:
This PR is 11 months old. Is it still relevant? Do we need to close it and open a new one later?

xincjin-NOAA commented:
> This PR is 11 months old. Is it still relevant? Do we need to close it and open a new one later?

@CoryMartin-NOAA This can be closed. We have a new one for this: 
https://github.com/NOAA-EMC/GDASApp/issues/1231

Thanks,
